src/mlpack/methods/reinforcement_learning/sac_impl.hpp

```diff
@@ -0,0 +1,332 @@
+  typename ReplayType
+>
+void SAC<
+  EnvironmentType,
+  QNetworkType,
+  PolicyNetworkType,
+  UpdaterType,
+  ReplayType
+>::Update()
+{
+  // Sample from previous experience.
+  arma::mat sampledStates;
+  std::vector<ActionType> sampledActions;
+  arma::rowvec sampledRewards;
+  arma::mat sampledNextStates;
+  arma::irowvec isTerminal;
+
+  replayMethod.Sample(sampledStates, sampledActions, sampledRewards,
+      sampledNextStates, isTerminal);
+
+  // Critic network update.
+
+  // Get the actions for sampled next states, from policy.
+  arma::mat nextStateActions;
+  policyNetwork.Predict(sampledNextStates, nextStateActions);
+
+  arma::mat targetQInput = arma::join_vert(nextStateActions,
+      sampledNextStates);
+  arma::rowvec Q1, Q2;
+  targetQ1Network.Predict(targetQInput, Q1);
+  // targetQ2Network.Predict(targetQInput, Q2);
+  arma::rowvec nextQ = sampledRewards +  config.Discount() * (1 - isTerminal)
+      % Q1;
+
+  arma::mat sampledActionValues(action.size, sampledActions.size());
+  for (size_t i = 0; i < sampledActions.size(); i++)
+    sampledActionValues.col(i) = sampledActions[i].action[0];
+  arma::mat learningQInput = arma::join_vert(sampledActionValues,
+      sampledStates);
+  learningQ1Network.Forward(learningQInput, Q1);
+  // learningQ2Network.Forward(learningQInput, Q2);
+
+  arma::mat gradQ1Loss, gradQ2Loss;
+  lossFunction.Backward(Q1, nextQ, gradQ1Loss);
+  // lossFunction.Backward(Q2, nextQ, gradQ2Loss);
+
+  // Update the critic networks.
+  arma::mat gradientQ1, gradientQ2;
+  learningQ1Network.Backward(learningQInput, gradQ1Loss, gradientQ1);
+  #if ENS_VERSION_MAJOR == 1
+  qNetworkUpdater.Update(learningQ1Network.Parameters(), config.StepSize(),
+      gradientQ1);
+  #else
+  qNetworkUpdatePolicy->Update(learningQ1Network.Parameters(),
+      config.StepSize(), gradientQ1);
+  #endif
+  // learningQ2Network.Backward(learningQInput, gradQ2Loss, gradientQ2);
+  // #if ENS_VERSION_MAJOR == 1
+  // qNetworkUpdater.Update(learningQ2Network.Parameters(), config.StepSize(),
+  //     gradientQ2);
+  // #else
+  // qNetworkUpdatePolicy->Update(learningQ2Network.Parameters(),
+  //     config.StepSize(), gradientQ2);
+  // #endif
+
+  // Actor network update.
+
+  // arma::rowvec pi;
+  // policyNetwork.Predict(sampledStates, pi);
+
+  // arma::mat qInput = arma::join_vert(pi, sampledStates);
+  // learningQ1Network.Predict(qInput, Q1);
+  // learningQ2Network.Predict(qInput, Q2);
+
+  arma::mat gradient;
+  for (size_t i = 0; i < sampledStates.n_cols; i++)
+  {
+    arma::mat grad, gradQ, q;
+    arma::colvec singleState = sampledStates.col(i);
+    arma::colvec singlePi;
+    policyNetwork.Forward(singleState, singlePi);
+
+    arma::colvec input = arma::join_vert(singlePi, singleState);
+    learningQ1Network.Forward(input, q);
+    learningQ1Network.Backward(input, -q, gradQ);
+
+    //! TODO: Check for the correct location of pi grad.
+    arma::mat gradPolicy = gradQ.rows(0, singlePi.n_rows - 1);
```

Comment:
As you can see here, I'm taking the gradient for the policy network, from the gradient for q network, by taking the first value(as `singlePi.n_rows` is `1`; because Pendulum has action_space = `1`) of the gradQ matrix(which essentially is a colvec, because `gradQ` is a mat with a single column).

I take the first value because I want the gradient corresponding to the `singlePi` of the input that I used in `learningQ1Network`.

I mention this because I'm a bit skeptical about it. Is this the correct way to get the gradient for a particular input(here `singlePi`?

Reply:
Yes, that should work, I'll have to take a closer look into the issue, you are right #2454 has most probably the same issue.

Reply:
On my system the test converges most of the time, I guess on your system it never converges?

Reply:
I think you must be using `-1250` as reward threshold for the pendulum environemnt, as in the test. But any random agent will be able to reach that, as the average reward for pendulum env is close to -1250.

I've set up the rewardThreshold as `-1250` in the test, just to ensure that while making changes in the agent or anywhere for that matter, we don't disrupt the feed forward flow of the SAC agent. Once the agent is able to converge to a better reward, I'll set a higher threshold. So for now, can u set a reward maybe around `-900` and try?

Reply:
Sure, let me do that.